hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java

```diff
@@ -604,4 +596,34 @@ public void setConf(Configuration conf) {
   public Configuration getConf() {
     return conf;
   }
+
+  @VisibleForTesting
+  HiveClientPool loadHiveClientPool(Configuration configuration) {
+    String metastoreUri = configuration.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+    Pair<HiveClientPool, Long> cacheEntry = CLIENT_POOL_CACHE.getIfPresent(metastoreUri);
+    HiveClientPool clientPool = cacheEntry == null ? new HiveClientPool(configuration) : cacheEntry.first();
+    CLIENT_POOL_CACHE.put(metastoreUri, Pair.of(clientPool, System.currentTimeMillis()));
+    return clientPool;
+  }
+
+  private void scheduleCacheCleaner() {
+    if (cleaner == null) {
+      synchronized (HiveCatalog.class) {
+        if (cleaner == null) {
+          cleaner = Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                          .setDaemon(true)
+                          .setNameFormat("iceberg-client-pool-cache-cleaner-%d")
+                          .build());
+        }
+        ScheduledFuture<?> futures = cleaner.scheduleWithFixedDelay(() -> CLIENT_POOL_CACHE.asMap().entrySet().stream()
+                .filter(e -> e.getValue().second() + evictionInterval <= System.currentTimeMillis())
+                .forEach(e -> {
+                  HiveClientPool pool = e.getValue().first();
+                  CLIENT_POOL_CACHE.invalidate(e.getKey());
+                  pool.close();
```

Comment:
Could we have some problem with concurrency?
We decide to evict, but just after the check someone gets this pool, but we close it...

Reply:
You are right, I missed this one. We should synchronize on the CLIENT_POOL_CACHE when inserting/deleting entries.